Add ResetProperties method to ConnectionContext interface

monich opened the pull request:
Allows to reset connection context properties back to default.

thp commented on ofono/src/gprs.c:
```diff
@@ -893,6 +895,140 @@ static DBusMessage *pri_get_properties(DBusConnection *conn,
 	return reply;
 }
 
+static gboolean pri_update_str(char *val, const char *newval)
```
This doesn't have any bounds check?

monich replied:
Bounds check is done by pri_valid_ap.

tkenakka commented on ofono/src/gprs.c:
```diff
@@ -893,6 +895,140 @@ static DBusMessage *pri_get_properties(DBusConnection *conn,
+	struct ofono_sim *sim = __ofono_atom_find(OFONO_ATOM_TYPE_SIM, modem);
+	struct ofono_gprs_provision_data *settings;
+	int i, count;
+
```
Please add 
	if (ctx->active == TRUE)
		return __ofono_error_in_use(msg);

monich replied:
I could implement re-provisioning entirely on the client side - just parse mbpi file, deactivate the context and apply the properties. In fact, I have already done that and I didn't like it at all, it's too complex and fragile (too many steps involved + race condition at every step). Doing it on the ofono side is much easier, more reliable and allows to avoid duplicating the logic of selecting settings from mbpi (like, provider name matching).

Forcing the client to disconnect first breaks the process in two steps which complicates it and makes it prone to race condition, i.e. client A (settings) disconnects the context, client B (connman) re-connects it back, client A fails to reset settings.

What about deactivating the context if anything significant (apn/username/password) has changed?

tkenakka commented:
@monich 
1) This is a core change so could you propose this to upstream and ask their comments, in case they are open to this in some form? We could take their hints even if they don't take this.
2) I'll discuss a few other items with you via other medium.

monich commented:
Updated this PR to deactivate the context prior to updating any significant property. Added reset of the protocol property which I originally forgot. Tested on device. Please take another look.

jpoutiai commented:
Shouldn't you also set ctx->pending = dbus_message_ref(msg); when deactivating context prior resetting or add pending_reset check at least to pri_set_property? Otherwise another client may also request deactivation and receive incorrect error as a reply.

monich commented:
Good point. Will fix.

tkenakka commented on ofono/src/gprs.c:
```diff
@@ -820,6 +832,252 @@ static void pri_update_mms_context_settings(struct pri_context *ctx)
+	for (i = 0; i < count; i++) {
+		const struct ofono_gprs_provision_data *ap = settings + i;
+		if (ap->type == ctx->type && ap_valid(ap)) {
+			if (!(ctx->active || ctx->pending) ||
```
Wouldn't following be more easy to comprehend, or is it just me?
if ((!ctx->active && !ctx->pending)

monich replied:
To me two logical operations NOT (x OR y) is easier to comprehend than three (NOT x) AND (NOT y)

monich commented:
Updated the PR to handle "deactivate while deactivating for reset" case. SetProperty or DeactivateAll will fail with org.ofono.Error.InProgress if the context is being deactivated for reset.

monich commented:
Superseded by https://github.com/nemomobile-packages/ofono/pull/282
